src/ethereum/prague/vm/instructions/environment.py

```diff
@@ -583,3 +605,135 @@ def blob_base_fee(evm: Evm) -> None:
 
     # PROGRAM COUNTER
     evm.pc += 1
+
+
+def returndataload(evm: Evm) -> None:
+    """
+    Copies data from the return data buffer code to memory
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    # STACK
+    offset = pop(evm.stack)
+
+    # GAS
+    charge_gas(evm, GAS_VERY_LOW)
+
+    # OPERATION
+    value = U256.from_be_bytes(buffer_read(evm.return_data, offset, U256(32)))
+    push(evm.stack, value)
+
+    # PROGRAM COUNTER
+    evm.pc += 1
+
+
+def dataload(evm: Evm) -> None:
+    """
+    Pushes 32-byte word to stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    # STACK
+    offset = pop(evm.stack)
+
+    # GAS
+    charge_gas(evm, GAS_DATALOAD)
+
+    # OPERATION
+    assert evm.eof is not None
+    value = U256.from_be_bytes(
+        buffer_read(evm.eof.metadata.data_section_contents, offset, U256(32))
+    )
+    push(evm.stack, value)
+
+    # PROGRAM COUNTER
+    evm.pc += 1
+
+
+def dataload_n(evm: Evm) -> None:
+    """
+    Pushes 32-byte word to stack where the word is addressed
+    by a static immediate argument
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    # STACK
+    pass
+
+    # GAS
+    charge_gas(evm, GAS_DATALOADN)
+
+    # OPERATION
+    assert evm.eof is not None
+    offset = U256.from_be_bytes(evm.code[evm.pc + 1 : evm.pc + 3])
+    value = U256.from_be_bytes(
+        buffer_read(evm.eof.metadata.data_section_contents, offset, U256(32))
+    )
+    push(evm.stack, value)
+
+    # PROGRAM COUNTER
+    # 1 + 2 bytes of immediate data
+    evm.pc += 3
+
+
+def datasize(evm: Evm) -> None:
+    """
+    Pushes the data section size.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    # STACK
+    pass
+
+    # GAS
+    charge_gas(evm, GAS_DATASIZE)
+
+    # OPERATION
+    assert evm.eof is not None
+    push(evm.stack, U256(len(evm.eof.metadata.data_section_contents)))
+
+    # PROGRAM COUNTER
+    evm.pc += 1
+
+
+def datacopy(evm: Evm) -> None:
+    """
+    Copies a segment of data section to memory.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+    """
+    # STACK
+    memory_start_index = pop(evm.stack)
+    offset = pop(evm.stack)
+    size = pop(evm.stack)
+
+    # GAS
+    copy_gas_cost = 3 + 3 * ((Uint(size) + 31) // 32)
+    extend_memory = calculate_gas_extend_memory(
+        evm.memory, [(memory_start_index, size)]
+    )
+    charge_gas(evm, copy_gas_cost + extend_memory.cost)
+
+    # OPERATION
+    assert evm.eof is not None
```

Comment:
I assume there's a check somewhere else that enforces `datacopy` is only used in EOF contracts? Should we leave a comment explaining that?